src/Microsoft.AspNet.OData.Shared/Query/Expressions/ExpressionBinderBase.cs

```diff
@@ -1176,17 +1176,29 @@ internal Expression CreatePropertyAccessExpression(Expression source, IEdmProper
             else
             {
                 return GetFlattenedPropertyExpression(propertyPath)
-                    ?? ConvertNonStandardPrimitives(ExpressionBinderBase.GetPropertyExpression(source, (this.HasInstancePropertyContainer && !propertyPath.Contains("\\") ? "Instance\\" : String.Empty) + propertyName));
+                    ?? ConvertNonStandardPrimitives(ExpressionBinderBase.GetPropertyExpression(source, (this.HasInstancePropertyContainer && !propertyPath.Contains("\\") ? "Instance\\" : String.Empty) + propertyName, property, Model));
             }
         }
 
-        internal static Expression GetPropertyExpression(Expression source, string propertyPath)
+        internal static Expression GetPropertyExpression(Expression source, string propertyPath, IEdmProperty property = null, IEdmModel model = null)
         {
             string[] propertyNameParts = propertyPath.Split('\\');
             Expression propertyValue = source;
-            foreach (var propertyName in propertyNameParts)
+            for (int i = 0; i < propertyNameParts.Length; i++)
             {
-                propertyValue = Expression.Property(propertyValue, propertyName);
+                // property access over declaring type from model without extra casts
+                // - solves property path expressions like Stamp/CreatedByUser/Name, where CreatedByUser is interface which do not directly have a Name property, but the Name is in some base interface up in the hierarchy
```

Comment:
When you talk about a "base interface up in the hierarchy", do you mean a type that inerhits from `CreatedByUser` or a type that `CreatedByUser` inherits from?

Why not use that type that has the property in the path expressions instead of using a type that is not guaranteed to have that property?

Could you share more information about concrete scenarios where this issue occurs and that this PR aims to fix?